core/src/main/scala/org/apache/spark/Accumulable.scala

```diff
@@ -146,6 +212,32 @@ class Accumulable[R, T] private (
   def merge(term: R) { value_ = param.addInPlace(value_, term)}
 
   /**
+   * Merge in pending updates for ac consistent accumulators or merge accumulated values for
+   * regular accumulators. This is only called on the driver when merging task results together.
+   */
+  private[spark] def internalMerge(term: Any) {
+    if (!consistent) {
+      merge(term.asInstanceOf[R])
+    } else {
+      mergePending(term.asInstanceOf[mutable.HashMap[(Int, Int, Int), R]])
+    }
+  }
+
+  /**
+   * Merge another Accumulable's pending updates, checks to make sure that each pending update has
+   * not already been processed before updating.
+   */
+  private[spark] def mergePending(term: mutable.HashMap[(Int, Int, Int), R]) = {
+    term.foreach{case ((rddId, shuffleId, splitId), v) =>
+      val splits = processed.getOrElseUpdate((rddId, shuffleId), new mutable.BitSet())
+      if (!splits.contains(splitId)) {
+        splits += splitId
+        value_ = param.addInPlace(value_, v)
+      }
```

Comment:
Sure we could do that - I'd kept the separate processed since I thought the space efficiency of a bitset might be worth it as well as it seemed like it might be more confusing to have one val with two different meanings between driver & worker.